### ISSUE TITLE
Protocol scoping

### ISSUE THREAD

**Issue by jakebeal** [SD2E/paml#15]
On the one hand, protocols need to be self-contained and it would be problematic if protocols could have elements mixed together across lexical scopes.

On the other hand, there are times that we will want to access interiors, such as exposing parameters in a sub-protocol for visibility and control.

To what degree do we want such interfaces to be pre-specified and rigid vs. permissive and able to be determined dynamically? How should the affect our representations?

For example, consider a protocol that is set up to culture at 37 degrees, and we want to run it culturing at 30 degrees instead. Ways this might be implemented include:
- Forking the protocol and making a version that runs at 30 degrees
- Forking the protocol and making a version in which temperature is an input with a default value
- Dynamically exposing the 37 degree value to make it an input with a default value
- Allowing a "patch" that forcibly overrides the 30 degree value

I tend to favor the less dynamic methods myself as being less messy, but we don't want things to be too rigid or else one gets a new type of mess in the development of a bazillion forks protocols.

**Comment by bjkeller** [comment]
Your example is really about the interface of the protocol. I would say allowing flexibility in interfaces is problematic. 

In principle, the input of an action should either have a default, be assigned a value by the activity, or be linked to a parameter of the activity.

An activity parameter may have a default. 
So, I would say in your example that making a new version that exposes the temperature as a parameter with a default makes sense.

**Comment by danbryce** [comment]
I think variables are related to this issue.  Variables are a way to provide flexibility to the interface and cause us to consider complications wrt. #14

**Comment by bjkeller** [comment]
I was hoping to exclude use of variables as a means to short-circuit object flow.

**Comment by jakebeal** [comment]
I think that object flow would be happening in all of these cases, and the difference is the degree to which it is predictable through static analysis.

**Comment by danbryce** [comment]
There are two types of input variables that we should distinguish:

1) Causal: variables whose assignment will be used as conditions for other variable assignments.  I.e., they have an impact on other variables.  I'm thinking mainly of branch conditions or primitive inputs that effect the outputs of the primitive.  The causal variable doesn't need to be dynamic, but we would require dynamic execution to compute the transitive closure of its causal influence.

2) Latent: anything that is not a causal variable.  These should be statically checkable.

We should be okay with including only latent variables.  I'm thinking of things like gain settings, temperature, etc.

**Comment by jakebeal** [comment]
Note that static checking over causal variables is possible as well (and frequently done in programming language analysis), but is harder and often does not allow as tight an inference.

**Comment by danbryce** [comment]
I agree, but I think the issue here is that you need to invoke library calls (i.e., primitives) that don’t have a defined semantics.  That is typically why you need to do dynamic analysis.  POML would provide that semantics so that you could do static analysis.  Is there more to it than that?

**Comment by jakebeal** [comment]
Consensus decision at 5/17 protocols retreat: strict lexical scoping.

Protocols cannot refer to or modify the contents of another protocol; they can only impact them when "invited" via protocol inputs (parameters) and outputs.
